File: src/android/app/src/main/kotlin/com/amirawellness/data/models/EmotionalTrend.kt
package com.amirawellness.data.models

import android.os.Parcelable // Android SDK
import kotlinx.parcelize.Parcelize // Kotlin Android Extensions
import com.google.gson.annotations.SerializedName // Gson 2.9.0
import java.util.Date // Java SDK
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import com.amirawellness.core.constants.AppConstants.EmotionType
import com.amirawellness.R

/**
 * Data model class representing emotional trend data for visualization and analysis in the Amira Wellness Android application.
 * This model captures patterns in emotional states over time, including trend direction, intensity changes, and data points for visualization.
 */

/**
 * Enumeration representing the direction of an emotional trend over time.
 */
enum class TrendDirection {
    INCREASING,
    DECREASING,
    STABLE,
    FLUCTUATING;

    /**
     * Returns the localized display name of the trend direction.
     *
     * @return Localized display name of the trend direction
     */
    fun getDisplayName(): String {
        return when (this) {
            INCREASING -> "Aumentando" // Increasing in Spanish
            DECREASING -> "Disminuyendo" // Decreasing in Spanish
            STABLE -> "Estable" // Stable in Spanish
            FLUCTUATING -> "Fluctuando" // Fluctuating in Spanish
        }
    }

    /**
     * Returns the localized description of the trend direction.
     *
     * @return Localized description of the trend direction
     */
    fun getDescription(): String {
        return when (this) {
            INCREASING -> "Esta emoción está aumentando con el tiempo"
            DECREASING -> "Esta emoción está disminuyendo con el tiempo"
            STABLE -> "Esta emoción se mantiene estable"
            FLUCTUATING -> "Esta emoción fluctúa significativamente"
        }
    }

    /**
     * Returns the resource ID for the icon associated with the trend direction.
     *
     * @return Resource ID for the trend direction icon
     */
    fun getIconResId(): Int {
        return when (this) {
            INCREASING -> R.drawable.ic_trend_increasing
            DECREASING -> R.drawable.ic_trend_decreasing
            STABLE -> R.drawable.ic_trend_stable
            FLUCTUATING -> R.drawable.ic_trend_fluctuating
        }
    }
}

/**
 * Enumeration representing the time period for trend analysis.
 */
enum class PeriodType {
    DAY,
    WEEK,
    MONTH;

    /**
     * Returns the localized display name of the period type.
     *
     * @return Localized display name of the period type
     */
    fun getDisplayName(): String {
        return when (this) {
            DAY -> "Hoy" // Today in Spanish
            WEEK -> "Esta semana" // This week in Spanish
            MONTH -> "Este mes" // This month in Spanish
        }
    }

    /**
     * Returns the default date range for the period type.
     *
     * @return Pair of start and end dates
     */
    fun getDateRange(): Pair<Date, Date> {
        val calendar = Calendar.getInstance()
        val endDate = calendar.time
        
        when (this) {
            DAY -> calendar.add(Calendar.DAY_OF_YEAR, -1)
            WEEK -> calendar.add(Calendar.WEEK_OF_YEAR, -1)
            MONTH -> calendar.add(Calendar.MONTH, -1)
        }
        
        val startDate = calendar.time
        return Pair(startDate, endDate)
    }
}

/**
 * Data class representing a single data point in an emotional trend.
 */
@Parcelize
data class TrendDataPoint(
    val date: Date,
    val value: Int,
    val context: String? = null
) : Parcelable {

    /**
     * Returns a formatted string representation of the date.
     *
     * @param formatter Optional formatter to use
     * @return Formatted date string
     */
    fun getFormattedDate(formatter: DateFormat? = null): String {
        val dateFormatter = formatter ?: SimpleDateFormat.getDateInstance(DateFormat.SHORT)
        return dateFormatter.format(date)
    }

    /**
     * Returns a formatted string representation of the value.
     *
     * @return Formatted value string (e.g., '7/10')
     */
    fun getFormattedValue(): String {
        return "$value/10"
    }
}

/**
 * Data class representing an insight derived from emotional trend analysis.
 */
@Parcelize
data class EmotionalInsight(
    val type: String,
    val description: String,
    val relatedEmotions: List<EmotionType>,
    val confidence: Double,
    val recommendedActions: List<String>
) : Parcelable {

    /**
     * Returns a formatted string representation of the confidence level.
     *
     * @return Formatted confidence string (e.g., '85%')
     */
    fun getFormattedConfidence(): String {
        return "${(confidence * 100).toInt()}%"
    }

    /**
     * Returns a comma-separated list of related emotion names.
     *
     * @return Comma-separated list of emotion names
     */
    fun getRelatedEmotionsText(): String {
        return relatedEmotions.joinToString(", ") { emotion ->
            when (emotion) {
                EmotionType.JOY -> "Alegría"
                EmotionType.SADNESS -> "Tristeza"
                EmotionType.ANGER -> "Ira"
                EmotionType.FEAR -> "Miedo"
                EmotionType.DISGUST -> "Disgusto"
                EmotionType.SURPRISE -> "Sorpresa"
                EmotionType.TRUST -> "Confianza"
                EmotionType.ANTICIPATION -> "Anticipación"
                EmotionType.ANXIETY -> "Ansiedad"
                EmotionType.CALM -> "Calma"
                else -> emotion.toString()
            }
        }
    }
}

/**
 * Data class representing a data point DTO for API communication.
 */
data class DataPointDto(
    @SerializedName("date") val date: Date,
    @SerializedName("value") val value: Int,
    @SerializedName("context") val context: String?
)

/**
 * Data class representing an emotional trend DTO for API communication.
 */
data class EmotionalTrendDto(
    @SerializedName("emotion_type") val emotionType: String,
    @SerializedName("data_points") val dataPoints: List<DataPointDto>,
    @SerializedName("overall_trend") val overallTrend: String,
    @SerializedName("average_intensity") val averageIntensity: Double,
    @SerializedName("peak_intensity") val peakIntensity: Double,
    @SerializedName("peak_date") val peakDate: Date,
    @SerializedName("occurrence_count") val occurrenceCount: Int
)

/**
 * Data class representing an emotional trend over time.
 */
@Parcelize
data class EmotionalTrend(
    val emotionType: EmotionType,
    val dataPoints: List<TrendDataPoint>,
    val overallTrend: TrendDirection,
    val averageIntensity: Double,
    val peakIntensity: Double,
    val peakDate: Date,
    val occurrenceCount: Int
) : Parcelable {

    /**
     * Returns a localized description of the emotional trend.
     *
     * @return Localized description of the trend
     */
    fun getTrendDescription(): String {
        val emotion = when (emotionType) {
            EmotionType.JOY -> "alegría"
            EmotionType.SADNESS -> "tristeza"
            EmotionType.ANGER -> "ira"
            EmotionType.FEAR -> "miedo"
            EmotionType.DISGUST -> "disgusto"
            EmotionType.SURPRISE -> "sorpresa"
            EmotionType.TRUST -> "confianza"
            EmotionType.ANTICIPATION -> "anticipación"
            EmotionType.ANXIETY -> "ansiedad"
            EmotionType.CALM -> "calma"
            else -> emotionType.toString().toLowerCase()
        }
        
        return "Tu nivel de $emotion ${overallTrend.getDescription().toLowerCase()} con una intensidad promedio de ${getFormattedAverageIntensity()}."
    }

    /**
     * Returns a formatted string representation of the average intensity.
     *
     * @return Formatted average intensity string (e.g., '7.2/10')
     */
    fun getFormattedAverageIntensity(): String {
        return String.format("%.1f/10", averageIntensity)
    }

    /**
     * Returns a formatted string representation of the peak intensity.
     *
     * @return Formatted peak intensity string (e.g., '9/10')
     */
    fun getFormattedPeakIntensity(): String {
        return String.format("%.1f/10", peakIntensity)
    }

    /**
     * Returns a formatted string representation of the peak date.
     *
     * @return Formatted peak date string
     */
    fun getFormattedPeakDate(): String {
        val dateFormat = SimpleDateFormat.getDateInstance(DateFormat.MEDIUM)
        return dateFormat.format(peakDate)
    }

    /**
     * Returns the date range covered by the trend data points.
     *
     * @return Pair with start and end dates
     */
    fun getDateRange(): Pair<Date, Date> {
        if (dataPoints.isEmpty()) {
            return Pair(Date(), Date())
        }
        
        val sortedPoints = dataPoints.sortedBy { it.date }
        return Pair(sortedPoints.first().date, sortedPoints.last().date)
    }

    /**
     * Returns the range of intensity values in the trend data points.
     *
     * @return Pair with minimum and maximum intensity values
     */
    fun getIntensityRange(): Pair<Int, Int> {
        if (dataPoints.isEmpty()) {
            return Pair(0, 10)
        }
        
        val minValue = dataPoints.minOf { it.value }
        val maxValue = dataPoints.maxOf { it.value }
        return Pair(minValue, maxValue)
    }

    /**
     * Returns a formatted string representation of the trend period.
     *
     * @return Formatted period string (e.g., 'Last 7 days')
     */
    fun getFormattedPeriod(): String {
        val (startDate, endDate) = getDateRange()
        val diffMillis = endDate.time - startDate.time
        val diffDays = diffMillis / (1000 * 60 * 60 * 24)
        
        return when {
            diffDays < 1 -> "Hoy"
            diffDays < 7 -> "Últimos ${diffDays.toInt()} días"
            diffDays < 30 -> "Últimas ${(diffDays / 7).toInt()} semanas"
            else -> "Últimos ${(diffDays / 30).toInt()} meses"
        }
    }

    /**
     * Converts the EmotionalTrend model to a DTO for API communication.
     *
     * @return DTO representation of this emotional trend
     */
    fun toEmotionalTrendDto(): EmotionalTrendDto {
        return EmotionalTrendDto(
            emotionType = emotionType.toString(),
            dataPoints = dataPoints.map { DataPointDto(it.date, it.value, it.context) },
            overallTrend = overallTrend.toString(),
            averageIntensity = averageIntensity,
            peakIntensity = peakIntensity,
            peakDate = peakDate,
            occurrenceCount = occurrenceCount
        )
    }

    companion object {
        /**
         * Creates an EmotionalTrend instance from a DTO.
         *
         * @param dto DTO representation of the emotional trend
         * @return Model instance created from the DTO
         */
        fun fromDto(dto: EmotionalTrendDto): EmotionalTrend {
            val emotionType = try {
                EmotionType.valueOf(dto.emotionType)
            } catch (e: IllegalArgumentException) {
                // Default to CALM if the emotion type is not recognized
                EmotionType.CALM
            }
            
            val dataPoints = dto.dataPoints.map { 
                TrendDataPoint(it.date, it.value, it.context) 
            }
            
            val overallTrend = try {
                TrendDirection.valueOf(dto.overallTrend)
            } catch (e: IllegalArgumentException) {
                // Default to STABLE if the trend direction is not recognized
                TrendDirection.STABLE
            }
            
            return EmotionalTrend(
                emotionType = emotionType,
                dataPoints = dataPoints,
                overallTrend = overallTrend,
                averageIntensity = dto.averageIntensity,
                peakIntensity = dto.peakIntensity,
                peakDate = dto.peakDate,
                occurrenceCount = dto.occurrenceCount
            )
        }

        /**
         * Calculates the trend direction based on a series of data points.
         *
         * @param dataPoints List of data points to analyze
         * @return Calculated trend direction
         */
        fun calculateTrendDirection(dataPoints: List<TrendDataPoint>): TrendDirection {
            if (dataPoints.size < 2) {
                return TrendDirection.STABLE
            }
            
            // Sort data points by date
            val sortedPoints = dataPoints.sortedBy { it.date }
            
            // Linear regression to find slope
            val n = sortedPoints.size
            val xValues = sortedPoints.mapIndexed { index, _ -> index.toDouble() }
            val yValues = sortedPoints.map { it.value.toDouble() }
            
            val sumX = xValues.sum()
            val sumY = yValues.sum()
            val sumXY = xValues.zip(yValues).sumOf { it.first * it.second }
            val sumXX = xValues.sumOf { it * it }
            
            val slope = if (n * sumXX - sumX * sumX != 0.0) {
                (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX)
            } else {
                0.0
            }
            
            // Calculate variance to determine stability
            val mean = yValues.average()
            val variance = yValues.sumOf { (it - mean) * (it - mean) } / n
            
            // Determine trend direction
            return when {
                variance < 1.0 -> TrendDirection.STABLE
                slope > 0.2 -> TrendDirection.INCREASING
                slope < -0.2 -> TrendDirection.DECREASING
                else -> TrendDirection.FLUCTUATING
            }
        }
    }
}

/**
 * Data class representing a request for emotional trend analysis.
 */
data class EmotionalTrendRequest(
    val periodType: PeriodType,
    val startDate: Date,
    val endDate: Date,
    val emotionTypes: List<EmotionType>? = null
) {

    /**
     * Validates that the request has valid parameters.
     *
     * @return True if the request is valid, false otherwise
     */
    fun isValid(): Boolean {
        // Check if startDate is before endDate
        if (startDate.after(endDate)) {
            return false
        }
        
        // Check if date range is within reasonable limits (e.g., not more than 1 year)
        val diffMillis = endDate.time - startDate.time
        val diffDays = diffMillis / (1000 * 60 * 60 * 24)
        if (diffDays > 365) {
            return false
        }
        
        return true
    }

    /**
     * Converts the request to a map of parameters for API requests.
     *
     * @return Map of parameters
     */
    fun toMap(): Map<String, Any> {
        val params = mutableMapOf<String, Any>()
        
        params["periodType"] = periodType.name
        params["startDate"] = formatDate(startDate)
        params["endDate"] = formatDate(endDate)
        
        emotionTypes?.let {
            params["emotionTypes"] = it.joinToString(",") { emotion -> emotion.name }
        }
        
        return params
    }
    
    private fun formatDate(date: Date): String {
        val formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        formatter.timeZone = TimeZone.getTimeZone("UTC")
        return formatter.format(date)
    }
}

/**
 * Data class representing a response from emotional trend analysis.
 */
data class EmotionalTrendResponse(
    val trends: List<EmotionalTrend>,
    val insights: List<EmotionalInsight>
)

File: src/ios/AmiraWellness/AmiraWellness/Models/EmotionalTrend.swift
import Foundation // For basic data types and JSON handling

// MARK: - TrendDirection

@frozen
enum TrendDirection: String, Codable {
    case increasing
    case decreasing
    case stable
    case fluctuating
    
    func displayName() -> String {
        switch self {
        case .increasing:
            return NSLocalizedString("En aumento", comment: "Increasing trend direction")
        case .decreasing:
            return NSLocalizedString("En descenso", comment: "Decreasing trend direction")
        case .stable:
            return NSLocalizedString("Estable", comment: "Stable trend direction")
        case .fluctuating:
            return NSLocalizedString("Fluctuante", comment: "Fluctuating trend direction")
        }
    }
    
    func description() -> String {
        switch self {
        case .increasing:
            return NSLocalizedString("Esta emoción ha estado aumentando con el tiempo", comment: "Increasing trend description")
        case .decreasing:
            return NSLocalizedString("Esta emoción ha estado disminuyendo con el tiempo", comment: "Decreasing trend description")
        case .stable:
            return NSLocalizedString("Esta emoción ha permanecido constante", comment: "Stable trend description")
        case .fluctuating:
            return NSLocalizedString("Esta emoción ha estado fluctuando considerablemente", comment: "Fluctuating trend description")
        }
    }
    
    func icon() -> String {
        switch self {
        case .increasing:
            return "arrow.up.right"
        case .decreasing:
            return "arrow.down.right"
        case .stable:
            return "arrow.right"
        case .fluctuating:
            return "waveform.path"
        }
    }
}

// MARK: - TrendPeriodType

@frozen
enum TrendPeriodType: String, Codable {
    case daily
    case weekly
    case monthly
    
    func displayName() -> String {
        switch self {
        case .daily:
            return NSLocalizedString("Diario", comment: "Daily period type")
        case .weekly:
            return NSLocalizedString("Semanal", comment: "Weekly period type")
        case .monthly:
            return NSLocalizedString("Mensual", comment: "Monthly period type")
        }
    }
    
    func dateFormatter() -> DateFormatter {
        let formatter = DateFormatter()
        
        switch self {
        case .daily:
            formatter.dateStyle = .short
            formatter.timeStyle = .short
        case .weekly:
            formatter.dateFormat = "EEEE, d MMM"
        case .monthly:
            formatter.dateFormat = "d MMM, yyyy"
        }
        
        formatter.locale = Locale(identifier: "es")
        return formatter
    }
    
    func defaultRange() -> (start: Date, end: Date) {
        let calendar = Calendar.current
        let end = Date()
        
        switch self {
        case .daily:
            let start = calendar.date(byAdding: .day, value: -1, to: end)!
            return (start, end)
        case .weekly:
            let start = calendar.date(byAdding: .day, value: -7, to: end)!
            return (start, end)
        case .monthly:
            let start = calendar.date(byAdding: .month, value: -1, to: end)!
            return (start, end)
        }
    }
}

// MARK: - InsightType

@frozen
enum InsightType: String, Codable {
    case pattern
    case trigger
    case improvement
    case correlation
    case recommendation
    
    func displayName() -> String {
        switch self {
        case .pattern:
            return NSLocalizedString("Patrón", comment: "Pattern insight type")
        case .trigger:
            return NSLocalizedString("Desencadenante", comment: "Trigger insight type")
        case .improvement:
            return NSLocalizedString("Mejora", comment: "Improvement insight type")
        case .correlation:
            return NSLocalizedString("Correlación", comment: "Correlation insight type")
        case .recommendation:
            return NSLocalizedString("Recomendación", comment: "Recommendation insight type")
        }
    }
    
    func icon() -> String {
        switch self {
        case .pattern:
            return "waveform.path.ecg"
        case .trigger:
            return "bolt.fill"
        case .improvement:
            return "chart.line.uptrend.xyaxis"
        case .correlation:
            return "link"
        case .recommendation:
            return "lightbulb.fill"
        }
    }
}

// MARK: - TrendDataPoint

struct TrendDataPoint: Codable, Equatable {
    let date: Date
    let value: Int
    let context: String?
    
    init(date: Date, value: Int, context: String? = nil) {
        self.date = date
        self.value = value
        self.context = context
    }
    
    func formattedDate(formatter: DateFormatter? = nil) -> String {
        let dateFormatter = formatter ?? {
            let df = DateFormatter()
            df.dateStyle = .short
            df.timeStyle = .short
            df.locale = Locale(identifier: "es")
            return df
        }()
        
        return dateFormatter.string(from: date)
    }
    
    func formattedValue() -> String {
        return "\(value)/10"
    }
}

// MARK: - EmotionalInsight

struct EmotionalInsight: Codable, Equatable {
    let type: InsightType
    let description: String
    let relatedEmotions: [EmotionType]
    let confidence: Double // 0.0 - 1.0
    let recommendedActions: [String]
    
    init(type: InsightType, description: String, relatedEmotions: [EmotionType], confidence: Double, recommendedActions: [String]) {
        self.type = type
        self.description = description
        self.relatedEmotions = relatedEmotions
        self.confidence = min(max(confidence, 0.0), 1.0) // Ensure between 0 and 1
        self.recommendedActions = recommendedActions
    }
    
    func formattedConfidence() -> String {
        let percentage = Int(confidence * 100)
        return "\(percentage)%"
    }
    
    func relatedEmotionsText() -> String {
        let emotionNames = relatedEmotions.map { $0.displayName() }
        return emotionNames.joined(separator: ", ")
    }
}

// MARK: - EmotionalTrend

struct EmotionalTrend: Codable, Equatable {
    let emotionType: EmotionType
    let dataPoints: [TrendDataPoint]
    let overallTrend: TrendDirection
    let averageIntensity: Double
    let peakIntensity: Double
    let peakDate: Date
    let occurrenceCount: Int
    
    init(emotionType: EmotionType, dataPoints: [TrendDataPoint], overallTrend: TrendDirection, averageIntensity: Double, peakIntensity: Double, peakDate: Date, occurrenceCount: Int) {
        self.emotionType = emotionType
        self.dataPoints = dataPoints
        self.overallTrend = overallTrend
        self.averageIntensity = averageIntensity
        self.peakIntensity = peakIntensity
        self.peakDate = peakDate
        self.occurrenceCount = occurrenceCount
    }
    
    func trendDescription() -> String {
        let emotionName = emotionType.displayName()
        let trendName = overallTrend.displayName().lowercased()
        let avgIntensity = String(format: "%.1f", averageIntensity)
        
        return NSLocalizedString(
            "\(emotionName) ha estado \(trendName) con una intensidad promedio de \(avgIntensity)/10",
            comment: "Trend description format"
        )
    }
    
    func formattedAverageIntensity() -> String {
        let formattedValue = String(format: "%.1f", averageIntensity)
        return "\(formattedValue)/10"
    }
    
    func formattedPeakIntensity() -> String {
        let formattedValue = String(format: "%.1f", peakIntensity)
        return "\(formattedValue)/10"
    }
    
    func formattedPeakDate() -> String {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        formatter.locale = Locale(identifier: "es")
        return formatter.string(from: peakDate)
    }
    
    func dateRange() -> (start: Date, end: Date) {
        guard !dataPoints.isEmpty else {
            // Default to current date if no data points
            let current = Date()
            return (current, current)
        }
        
        let sortedDates = dataPoints.map { $0.date }.sorted()
        return (sortedDates.first!, sortedDates.last!)
    }
    
    func intensityRange() -> (min: Int, max: Int) {
        guard !dataPoints.isEmpty else {
            return (0, 0)
        }
        
        let values = dataPoints.map { $0.value }
        return (values.min()!, values.max()!)
    }
}

// MARK: - EmotionalTrendRequest

struct EmotionalTrendRequest: Codable, Equatable {
    let periodType: TrendPeriodType
    let startDate: Date
    let endDate: Date
    let emotionTypes: [EmotionType]?
    
    init(periodType: TrendPeriodType, startDate: Date, endDate: Date, emotionTypes: [EmotionType]? = nil) {
        self.periodType = periodType
        self.startDate = startDate
        self.endDate = endDate
        self.emotionTypes = emotionTypes
    }
    
    func isValid() -> Bool {
        // Ensure startDate is before endDate
        if startDate >= endDate {
            return false
        }
        
        // Ensure date range is not unreasonably large
        let calendar = Calendar.current
        let components = calendar.dateComponents([.day], from: startDate, to: endDate)
        if let days = components.day, days > 365 {
            return false
        }
        
        return true
    }
    
    func toParameters() -> [String: Any] {
        var parameters: [String: Any] = [:]
        
        parameters["periodType"] = periodType.rawValue
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        dateFormatter.timeZone = TimeZone(identifier: "UTC")
        
        parameters["startDate"] = dateFormatter.string(from: startDate)
        parameters["endDate"] = dateFormatter.string(from: endDate)
        
        if let types = emotionTypes {
            parameters["emotionTypes"] = types.map { $0.rawValue }.joined(separator: ",")
        }
        
        return parameters
    }
}

// MARK: - EmotionalTrendResponse

struct EmotionalTrendResponse: Codable, Equatable {
    let trends: [EmotionalTrend]
    let insights: [EmotionalInsight]
    
    init(trends: [EmotionalTrend], insights: [EmotionalInsight]) {
        self.trends = trends
        self.insights = insights
    }
}